Source/DafnyLanguageServer/Workspace/DocumentDatabase.cs

```diff
@@ -54,42 +56,50 @@ ISymbolTableRelocator symbolTableRelocator
     public async Task<bool> CloseDocumentAsync(TextDocumentIdentifier documentId) {
       if (documents.Remove(documentId.Uri, out var databaseEntry)) {
         databaseEntry.CancelPendingUpdates();
-        await databaseEntry.Document;
+        try {
+          await databaseEntry.VerifiedDocument;
+        } catch (TaskCanceledException) {
+        }
         return true;
       }
       return false;
     }
 
-    public async Task<DafnyDocument> OpenDocumentAsync(TextDocumentItem document) {
+    public IObservable<DafnyDocument> OpenDocument(TextDocumentItem document) {
       var cancellationSource = new CancellationTokenSource();
+      var resolvedDocument = OpenAsync(document, cancellationSource.Token);
+      var verifiedDocument = VerifyAsync(resolvedDocument, VerifyOnOpen, cancellationSource.Token);
       var databaseEntry = new DocumentEntry(
         document.Version,
-        Task.Run(() => OpenAndVerifyAsync(document, cancellationSource.Token)),
+        resolvedDocument,
+        verifiedDocument,
         cancellationSource
       );
       documents.Add(document.Uri, databaseEntry);
-      return await databaseEntry.Document;
+      return resolvedDocument.ToObservable().Where(d => !d.LoadCanceled).Concat(verifiedDocument.ToObservable());
     }
 
-    private async Task<DafnyDocument> OpenAndVerifyAsync(TextDocumentItem textDocument, CancellationToken cancellationToken) {
+    private async Task<DafnyDocument> OpenAsync(TextDocumentItem textDocument, CancellationToken cancellationToken) {
       try {
-        return await LoadAndVerifyAsync(textDocument, VerifyOnOpen, cancellationToken);
+        return await documentLoader.LoadAsync(textDocument, cancellationToken);
       } catch (OperationCanceledException) {
         // We do not allow canceling the load of the placeholder document. Otherwise, other components
         // start to have to check for nullability in later stages such as change request processors.
         return documentLoader.CreateUnloaded(textDocument, CancellationToken.None);
       }
     }
 
-    private async Task<DafnyDocument> LoadAndVerifyAsync(TextDocumentItem textDocument, bool verify, CancellationToken cancellationToken) {
-      var document = await documentLoader.LoadAsync(textDocument, cancellationToken);
-      if (!verify || document.Errors.HasErrors) {
-        return document;
+    private async Task<DafnyDocument> VerifyAsync(Task<DafnyDocument> documentTask, bool verify, CancellationToken cancellationToken) {
+#pragma warning disable VSTHRD003
```

Comment:
> I think @cpitclaudel has a point though. If a change occurs before and after a border of a diagnostic, that position of the diagnostic won't be relevant anymore and will land in new text.
Hence, I too believe you should move the boundary of the diagnostic, while it's temporary, so that it does not spill out in the new text.

I'm not sure what you mean by "If a change occurs before and after a border of a diagnostic". Do you mean the text deleted by the change covers the entire range of the diagnostic?

Reply:
> Could you remove these warning removal directives by awaiting for the document before calling VerifyAsync ? It is not catching exceptions anyway.

I can't because the caller is not `async`.